tools/sklearn/pca.py

```diff
@@ -0,0 +1,96 @@
+                        help='Kernel coefficient for rbf, poly and sigmoid kernels. Ignored by other kernels')
+    parser.add_argument('-tol', '--tolerance', type=float, default=0.0,
+                        help='Convergence tolerance for arpack. If 0, optimal value will be chosen by arpack')
+    parser.add_argument('-mi', '--max_iter', nargs='?', type=int, default=None,\
+                        help="Maximum number of iterations for arpack")
+    parser.add_argument('-d', '--degree', type=int, default=3,\
+                        help="Degree for poly kernels. Ignored by other kernels")
+    parser.add_argument('-cf', '--coef0', type=float, default=1.0,
+                        help='Independent term in poly and sigmoid kernels')
+    parser.add_argument('-e', '--eigen_solver', type=str.lower, default='auto', choices=['auto', 'dense', 'arpack'],
+                        help="Choose the type of eigen solver.")
+    parser.add_argument('-o', '--outfile',
+                        help="Base name for output file (no extension).")
+    args = parser.parse_args()
+
+    usecols = None
+    cols = []
+    pca_params = {}
+
+    if args.columns != 'all':
+        with open(args.infile) as f:
+            num_cols = len(f.readline().split(' '))
+
+        for i in args.column_indices.split(' '):
+            if int(i) > num_cols - 1:
+                raise ValueError('Column index specified is greater than total number of columns available')
+            cols.append(int(i))
+
+        if args.columns == "exclude":
+            usecols = sorted(set(range(num_cols)) - set(cols))
+        elif args.columns == "include":
+            usecols = sorted(set(cols))
+
+    pca_input = np.loadtxt(fname=args.infile, skiprows=int(args.header), usecols=usecols)
+
+    pca_params.update({'n_components': args.number})
+
+    if args.pca_type == 'classical':
+        pca_params.update({'svd_solver': args.svd_solver, 'whiten': args.whiten})
+        if args.svd_solver == 'arpack':
+            pca_params.update({'tol': args.tolerance})
+        pca = PCA()
+
+    elif args.pca_type == 'incremental':
+        pca_params.update({'batch_size': args.batch_size, 'whiten': args.whiten})
+        pca = IncrementalPCA()
+
+    elif args.pca_type == 'kernel':
+        pca_params.update({'kernel': args.kernel, 'eigen_solver': args.eigen_solver, 'gamma': args.gamma})
+
+        if args.kernel == 'poly':
+            pca_params.update({'degree': args.degree, 'coef0': args.coef0})
+        elif args.kernel == 'sigmoid':
+            pca_params.update({'coef0': args.coef0})
+        elif args.kernel == 'precomputed':
+            pca_input = np.dot(pca_input, pca_input.T)
+
+        if args.eigen_solver == 'arpack':
+            pca_params.update({'tol': args.tolerance, 'max_iter': args.max_iter})
+
+        pca = KernelPCA()
+
+    pca.set_params(**pca_params)
+    pca_output = pca.fit_transform(pca_input)
+    np.savetxt(fname=args.outfile, X=pca_output, fmt='%.4f')
```

Comment:
```suggestion
    np.savetxt(fname=args.outfile, X=pca_output, delimiter='\t')
```
@lorrainealisha75, can you use a tab as the delimiter? Galaxy expects it for tabular files and hopefully then the has_n_columns test will pass.